src/asset_library.rs

```diff
@@ -19,24 +19,180 @@
 //! and thus should be defined in clients of this crate.
+use arbitrary::{Arbitrary, Unstructured};
+use ark_serialize::*;
+use espresso_macros::ser_test;
+use image::{imageops, ImageFormat, ImageResult, RgbImage};
 use jf_cap::{
     keys::AuditorPubKey,
     structs::{AssetCode, AssetCodeSeed, AssetDefinition},
     BaseField, KeyPair, Signature, VerKey,
 };
+use jf_utils::tagged_blob;
 use serde::{Deserialize, Serialize};
 use std::collections::{HashMap, HashSet};
 use std::fmt::{self, Display, Formatter};
+use std::io::{BufRead, Seek};
 use std::iter::FromIterator;
 use std::ops::Index;
 use std::str::FromStr;
 use tagged_base64::TaggedBase64;
 
+const ICON_WIDTH: u32 = 64;
+const ICON_HEIGHT: u32 = 64;
+
+/// A small icon to display with an asset in a GUI interface.
+#[ser_test(arbitrary)]
+#[tagged_blob("ICON")]
+#[derive(Clone, Debug, PartialEq)]
+pub struct Icon(RgbImage);
+
+impl CanonicalSerialize for Icon {
+    fn serialize<W: Write>(&self, mut writer: W) -> Result<(), SerializationError> {
+        // The serialization format we will use is very simple: write the width as a u32, the height
+        // as a u32, and then the raw data.
+        writer
+            .write_all(&self.width().to_le_bytes())
+            .map_err(SerializationError::IoError)?;
+        writer
+            .write_all(&self.height().to_le_bytes())
+            .map_err(SerializationError::IoError)?;
+        writer
+            .write_all(self.0.as_raw())
+            .map_err(SerializationError::IoError)
+    }
+
+    fn serialized_size(&self) -> usize {
+        // 8 bytes for the width and height (4 bytes each) plus 3 bytes for each RGB pixel.
+        8 + 3 * self.0.as_raw().len()
+    }
+}
+
+impl CanonicalDeserialize for Icon {
+    fn deserialize<R: Read>(mut reader: R) -> Result<Self, SerializationError> {
+        // Read the width and height so we can figure out how much raw data there is.
+        let mut width_buf = [0; 4];
+        let mut height_buf = [0; 4];
+        reader
+            .read_exact(&mut width_buf)
+            .map_err(SerializationError::IoError)?;
+        reader
+            .read_exact(&mut height_buf)
+            .map_err(SerializationError::IoError)?;
+        // The raw buffer has size 3*width*height, since each RGB pixel takes 3 bytes.
+        let width = u32::from_le_bytes(width_buf);
+        let height = u32::from_le_bytes(height_buf);
+        let mut image_buf = vec![0; (width as usize) * (height as usize) * 3];
+        reader
+            .read_exact(&mut image_buf)
+            .map_err(SerializationError::IoError)?;
+        Ok(Self(RgbImage::from_raw(width, height, image_buf).unwrap()))
+    }
+}
+
+impl<'a> Arbitrary<'a> for Icon {
+    fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        let width = u.arbitrary_len::<[u8; 3]>()?;
+        let height = u.arbitrary_len::<Vec<[u8; 3]>>()?;
```

Comment:
`Vec<[u8; 3]>` -> `[u8; 3]`.

Reply:
I think this is correct as is, the height of the image corresponds to the length of a container whose elements (rows) are arrays of pixels, i.e. `Vec`s of `[u8; 3]`. I see how this is confusing, though, I'll add a comment